ocaml/typing/jkind.ml

```diff
@@ -374,16 +450,34 @@ module Const = struct
     let float32 =
       { jkind = mode_crossing Layout.Const.float32; name = "float32" }
 
-    let word = { jkind = not_mode_crossing Layout.Const.word; name = "word" }
+    let word =
+      let jkind =
+        { (not_mode_crossing Layout.Const.word) with
+          nullability_upper_bound = Non_null
```

Comment:
`word` must be `Non_null`. We can't have `nativeint#` have a jkind that's `Or_null`, because then `nativeint# array` would be impossible. Same for `bits32`, `bit64`, `float32`, `float64`.

Reply:
Something's wrong in the design then. I think the problem in the design is that we're trying to work with `array` at all.

New proposal: restrict `array` to work only with values, but allow `Uniform_array.t` to work with arbitrary layouts. (Not sure necessarily how to implement this given that `Uniform_array.t` is built on top of `array`, but design first, implement second.) I will raise this point internally.

Reply:
The separability axis should solve this long-term, as discussed.